File: string/kmp/kmp.go
package kmp

// 构造前缀表next
// i,j分别指向前缀和后缀的末尾，此外j还表示当前位置前缀后缀的最大公共长度
func GetNext(needle string) []int {
	next := make([]int, len(needle))
	j := 0
	next[0] = j

	for i := 1; i < len(needle); i++ {
		// 前缀和后缀有元素不相等了，需要向前回退，这个回退是一个连续的过程
		for j > 0 && needle[j] != needle[i] {
			j = next[j-1]
		}
		if needle[i] == needle[j] {
			j++
		}
		next[i] = j
	}
	return next
}

func StrStr(haystack string, needle string) int {
	n := len(needle)
	if n == 0 {
		return 0
	}

	j := 0
	next := GetNext(needle)
	for i := 0; i < len(haystack); i++ {
		for j > 0 && haystack[i] != needle[j] {
			j = next[j-1]
		}
		if haystack[i] == needle[j] {
			j++
		}
		if j == n {
			return i - n + 1
		}
	}
	return -1
}

File: string/kmp/kmp.java
package string.kmp;

public class KMP {
    // next[i]表示的是[0..i]这段字符串的最长公共的前缀后缀
    // 前缀：以首位开头，不包含末尾的所有子字符串
    // 后缀：以末尾结尾，不包含首位的所有子字符串
    private int[] next;
    private String pattern; 

    public KMP(String pattern) {
        this.pattern = pattern;
        // dp数组只和pattern有关系
        this.next = new int[pattern.length()];
        int j = 0;  // 用于指向前缀的末尾
        next[0] = j;
        for (int i = 1; i < pattern.length(); i++) {
            // 注意看，这里的i有了一个特殊的含义，它指向的是后缀的末尾
            while (j > 0 && pattern.charAt(j) != pattern.charAt(i)) {
                // 后缀和前缀的字符不相等了，就往前看一下，这里注意next数组的意义，最长公共前后缀
                j = next[j - 1];
            }

            if (pattern.charAt(j) == pattern.charAt(i)) {
                j++;
            }

            next[i] = j;
        }
    }

    // 借助dp数组，找到pattern在text中首次出现的位置
    public int search(String text) {
        int j = 0;
        for (int i = 0; i < text.length(); i++) {
            while (j > 0 && pattern.charAt(j) != text.charAt(i)) {
                j = next[j - 1];
            }
            if (pattern.charAt(j) == text.charAt(i)) {
                j++;
            }

            if (j == pattern.length()) {
                return i - pattern.length() + 1;
            }
        }

        return -1;
    }
}
